src/core/python2js.c

```diff
@@ -190,89 +189,128 @@ _python2js_bytes(PyObject* x)
   return hiwire_bytes(x_buff, length);
 }
 
+/** WARNING: This function is not suitable for fallbacks. If this function
+ * returns NULL, we must assume that the cache has been corrupted and bail out.
+ */
 static JsRef
 _python2js_sequence(PyObject* x, PyObject* map, int depth)
 {
-  JsRef jsarray = hiwire_array();
-  if (_python2js_add_to_cache(map, x, jsarray)) {
-    hiwire_decref(jsarray);
-    return NULL;
-  }
-  size_t length = PySequence_Size(x);
-  for (size_t i = 0; i < length; ++i) {
+  bool success = false;
+  PyObject* pyitem = NULL;
+  JsRef jsitem = NULL;
+  // result:
+  JsRef jsarray = NULL;
+
+  jsarray = hiwire_array();
+  FAIL_IF_MINUS_ONE(_python2js_add_to_cache(map, x, jsarray));
+  Py_ssize_t length = PySequence_Size(x);
+  FAIL_IF_MINUS_ONE(length);
+  for (Py_ssize_t i = 0; i < length; ++i) {
     PyObject* pyitem = PySequence_GetItem(x, i);
-    if (pyitem == NULL) {
-      // If something goes wrong converting the sequence (as is the case with
-      // Pandas data frames), fallback to the Python object proxy
-      _python2js_remove_from_cache(map, x);
-      hiwire_decref(jsarray);
-      PyErr_Clear();
-      Py_INCREF(x);
-      return pyproxy_new(x);
-    }
-    JsRef jsitem = _python2js_cache(pyitem, map, depth);
-    if (jsitem == NULL) {
-      _python2js_remove_from_cache(map, x);
-      Py_DECREF(pyitem);
-      hiwire_decref(jsarray);
-      return NULL;
-    }
-    Py_DECREF(pyitem);
+    FAIL_IF_NULL(pyitem);
+    FAIL_IF_NULL(jsitem = _python2js_cache(pyitem, map, depth));
```

Comment:
I know it works in C, but still it's maybe more readable with,

```suggestion
    jsitem = _python2js_cache(pyitem, map, depth);
    FAIL_IF_NULL(jsitem);
```

Reply:
Okay, will keep with that convention, was just experimenting a bit here.